posts/kotlin-like-java-but-better-1.md

```diff
@@ -0,0 +1,184 @@
+---
+authors:
+- Davis Mohar
+tags:
+- Java
+- Kotlin
+- Spring Boot
+date: 2015-04-23T12:21:50.000Z
+title: "Kotlin: Like Java, But Better (Part 1)"
+image: https://github.com/davismohar/blog-usa/blob/master/images/2021/05/kotlin-like-java-but-better-1.png
+---
+
+Kotlin has quickly risen to popularity as an alternative language to Java.
+In this blog series, I will will first provide a crash-course in Kotlin 
+examining some of the design differences between Kotlin and Java, and then add
+some Kotlin code to an existing Java Spring Boot microservice in part two.
+
+# Kotlin 101
+![Kotlin 101](https://github.com/davismohar/blog-usa/blob/master/images/2021/05/kotlin-like-java-but-better-2.png)
+
+Kotlin was introduced by JetBrains in 2011 as a new language for the JVM. It
+was designed to be fully interoperable with Java, and in several cases, has 
+replaced it. In 2019, Google announced that Kotlin was now the official 
+language for Android development. Tools such as Spring Boot and 
+[JHipster](https://www.jhipster.tech/) allow you to start from scratch with
+Kotlin, but you can also seamlessly add Kotlin to an existing Java codebase!
+
+# What Kotlin Does Differently
+## Variable Declaration
+There are two keywords to declare a variable: `var` and `val`. 
+
+`var` is used to declare a variable that behaves just as you would expect an
+ordinary Java variable to act. The type of the variable is declared after the
+variable name, and declaring the type is optional if the variable is 
+instantiated. 
+```kotlin
+var num = 12
+var name: String
+name = "Jane Doe"
+```
+
+`val` is used to declare variables as read-only, similar to declaring a 
+variable final in Java. Attempting to change the reference of a val will result
+in a compilation error.
+```kotlin
+val num = 12 
+num = 13 //compilation error
+```
+
+You also may notice the lack of semicolons. Fear not- I haven't made a mistake!
+Semicolons are optional in Kotlin, and you shouldn't ever need them. The 
+compiler will infer your statement endings.
+
+## Null Safety
+One of the key design decisions that the Kotlin team made was to enforce 
+[null safety](https://kotlinlang.org/docs/null-safety.html) at compile-time.
+All variables must explicitly be declared as either nullable or non-nullable. 
+```kotlin
+var foo: Int? // nullable Int variable
+var bar: Int // non-nullable Int
+
+bar = Null // compilation error
+```
+As a result of this feature, NullPointerExceptions will never be thrown in 
+pure Kotlin code unless you explicitly throw one. They won't be missed! Kotlin
+also provides some additional syntax to enable working with these new null 
+constraints. 
+
+Inline null-checks can be used to access methods or fields of an object that
+may or may not be null. If the object is null, then the expression is 
+evaluated as null. If the object is not null, then the expression is evaluated
+as normal.
```

Comment:
If the object is not null, then the expression is evaluated as `normal`.

Reply:
I'm leaving this for now. I feel like saying `normal` implies there is a language keyword called `normal`. Should I change the wording to "the expression is evaluated normally"?

Reply:
Ahhh, I see what you're saying.  Disregard!